ggml-qnn.cpp

```diff
@@ -0,0 +1,3196 @@
+}
+
+
+static ggml_backend_t ggml_backend_qnn_reg_init(const char * params, void * user_data) {
+    if (nullptr == params) {
+        //QNN library path
+        //can be hardcoded to "/data/local/tmp/" for Android command line application
+        //or specified in JNI layer for Android APK
+        params = "/data/local/tmp/";
+    }
+    ggml_backend_t qnn_backend = ggml_backend_qnn_init((int) (intptr_t) user_data, params);
+
+    return qnn_backend;
+}
+
+
+bool ggml_backend_is_qnn(ggml_backend_t backend) {
+    return backend != nullptr && ggml_guid_matches(backend->guid, ggml_backend_qnn_guid());
+}
+
+
+void ggml_backend_qnn_set_n_threads(ggml_backend_t backend, int n_threads) {
+    GGML_ASSERT(ggml_backend_is_qnn(backend));
+
+    struct ggml_backend_qnn_context * ctx = (struct ggml_backend_qnn_context *)backend->context;
+    ctx->threads = n_threads;
+}
+
+
+const char * ggml_backend_qnn_get_name(ggml_backend_t backend) {
+    return backend->iface.get_name(backend);
+}
+
+
+int ggml_backend_qnn_get_device_count() {
+    return GGML_QNN_MAX_DEVICES;
+}
+
+
+void ggml_backend_qnn_get_device_description(size_t dev_num, char * description, size_t description_size) {
+    if (nullptr == description || 0 == description_size) {
+        QNN_LOG_WARN("invalid param");
+        return;
+    }
+
+    if (dev_num >= GGML_QNN_MAX_DEVICES) {
+        QNN_LOG_WARN("invalid param");
+        return;
+    }
+
+    snprintf(description, description_size, "%s", g_qnn_mgr[dev_num].name);
+    QNN_LOG_DEBUG("description:%s", description);
+}
+
+
+ggml_backend_buffer_type_t ggml_backend_qnn_buffer_type(size_t device) {
+    if (device >= GGML_QNN_MAX_DEVICES) {
+        QNN_LOG_DEBUG("ggml_backend_qnn_buffer_type error: device_index:%d is out of range [0, %d]\n",
+               device, GGML_QNN_MAX_DEVICES - 1);
+        return nullptr;
+    }
+
+    static ggml_backend_buffer_type ggml_backend_qnn_buffer_types[GGML_QNN_MAX_DEVICES];
+
+    static bool ggml_backend_qnn_buffer_type_initialized = false;
+
+    if (!ggml_backend_qnn_buffer_type_initialized) {
+        for (int i = 0; i < GGML_QNN_MAX_DEVICES; i++) {
+            ggml_backend_qnn_buffer_types[i] = {
+                /* .iface   = */ {
+                    /* .get_name         = */ ggml_backend_qnn_buffer_type_name,
+                    /* .alloc_buffer     = */ ggml_backend_qnn_buffer_type_alloc_buffer,
+                    /* .get_alignment    = */ ggml_backend_qnn_buffer_type_get_alignment,
+                    /* .get_max_size     = */ ggml_backend_qnn_buffer_type_get_max_size,
+                    /* .get_alloc_size   = */ nullptr,// defaults to ggml_nbytes
+                    /* .supports_backend = */ ggml_backend_qnn_buffer_type_supports_backend,
+                    /* .is_host          = */ ggml_backend_qnn_buffer_is_host
+                },
+                /* .context = */ new ggml_backend_qnn_buffer_type_context { device, GGML_QNN_NAME + std::to_string(device) },
```

Comment:
sorry to reopen the dialog here, one thing forgot to say:

maybe its worth to have a static array to store those `ggml_backend_qnn_buffer_type_context`, then the lifespan of context will match the types iself perfectly.

Reply:
there are already a static arrayto store those ggml_backend_qnn_buffer_type_context:
```
ggml_backend_buffer_type_t ggml_backend_qnn_buffer_type(size_t device) {
    if (device >= GGML_QNN_MAX_DEVICES) {
        QNN_LOG_DEBUG("ggml_backend_qnn_buffer_type error: device_index:%d is out of range [0, %d]\n",
               device, GGML_QNN_MAX_DEVICES - 1);
        return nullptr;
    }    

    static ggml_backend_buffer_type ggml_backend_qnn_buffer_types[GGML_QNN_MAX_DEVICES];

    static bool ggml_backend_qnn_buffer_type_initialized = false;

    if (!ggml_backend_qnn_buffer_type_initialized) {
        for (int i = 0; i < GGML_QNN_MAX_DEVICES; i++) {
            ggml_backend_qnn_buffer_types[i] = {
                /* .iface   = */ {
                    /* .get_name         = */ ggml_backend_qnn_buffer_type_name,
                    /* .alloc_buffer     = */ ggml_backend_qnn_buffer_type_alloc_buffer,
                    /* .get_alignment    = */ ggml_backend_qnn_buffer_type_get_alignment,
                    /* .get_max_size     = */ ggml_backend_qnn_buffer_type_get_max_size,
                    /* .get_alloc_size   = */ nullptr,// defaults to ggml_nbytes
                    /* .supports_backend = */ ggml_backend_qnn_buffer_type_supports_backend,
                    /* .is_host          = */ ggml_backend_qnn_buffer_is_host
                },
                /* .context = */ new ggml_backend_qnn_buffer_type_context { device, GGML_QNN_NAME + std::to_string(device) },
            };
        }
        ggml_backend_qnn_buffer_type_initialized = true;
    }    

    return &ggml_backend_qnn_buffer_types[device];
}
```

I think we don't need care what's the implementation here(just follow the style in other existing ggml backends).

Reply:
hmmm, My primary concern is that using 'new' without a corresponding 'delete' could potentially interfere with runtime sanitizers like ASan. Making it static should resolve this issue.

Reply:
@zhouwg create a PR about the change here at your fork, ptal:
https://github.com/zhouwg/llama.cpp/pull/1

Reply:
you are PR in my forked project doesn't make sense. pls don't spent efforts in these language details and focus on the keypoint in this PR. thanks for your understanding.

Reply:
if you are correct, I can't believe there are memory leaks in original ggml backend subsystem. 

I just feel that you spent too much time/efforts on these language details. pls help to focus on the "real keypoint" in this PR. thanks for you understanding.

Reply:
- Firstly, I apologize for the issues caused by my PR on your branch which resulted in the app crashing. I will investigate and rectify this issue.
- I also want to apologize for my previous focus on language details. However, regarding the use of new without a corresponding delete, my concern is that it could potentially interfere with runtime sanitizers like ASan. This could increase the debugging effort for other teams.
- I'd like to discuss the data structure of the backend, using ggml_backend_buffer_type as an example:
  - `ggml_backend_buffer_type_i iface;` - This field contains the base operations of the buffer type, serving as an interface in object-oriented programming. If you're familiar with the C++ class memory layout, it's akin to the vtable and can be overridden by a derived class.
  - `ggml_backend_buffer_type_context_t context;` - This is used by the derived class to maintain its private data structure.
  - These two structures combine to form an object, demonstrating the object model in llama.cpp. This technique is commonly used in other well-known libraries to simulate object-oriented programming.
- Returning to the backend classes, correct me if I'm wrong, but there are three major classes:
  - `ggml_backend_buffer_type` - This acts more like a buffer manager. We bind it to a specific backend object so it can manage the used/available memory of each device, and then allocate buffers from them.
  - `ggml_backend_buffer` - This serves as the back buffer of a tensor, holding its internal data.
  - `ggml_backend` - In this PR, it represents each backend device, dispatching the computation request to real devices. Each device has its own device memory, so there's a ggml_backend_buffer_type object associated with it.

Reply:
1. pls focus on the real keypoint in this PR.
2. the cost is acceptable even there really has memory leak which introduced by your review suggestion and my modifications(my previous code no memory leak because my original ctx is NULL which follow the style in metal backend:
![Screenshot from 2024-06-08 23-15-41](https://github.com/ggerganov/llama.cpp/assets/6889919/c6259f08-ccd0-48bb-827c-4165e13d8925)

4.